auth/api/iam/s2s_vptoken.go

```
@@ -19,49 +19,180 @@
+			InternalError: err,
+		}
+	}
+
+	// Unmarshal presentation submission
+	var submission pe.PresentationSubmission
+	submissionDecoded, err := url.QueryUnescape(submissionEncoded)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: invalid query escaping",
+			InternalError: err,
+		}
+	}
+	if err = json.Unmarshal([]byte(submissionDecoded), &submission); err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: invalid JSON",
+			InternalError: err,
+		}
+	}
 
-func (s serviceToService) validateVPToken(params map[string]string) (string, error) {
-	submission := params["presentation_submission"]
-	scope := params["scope"]
-	vp_token := params["vp_token"]
-	if submission == "" || scope == "" || vp_token == "" {
-		// TODO: right error response
-		return "", errors.New("missing required parameters")
-	}
-	// TODO: https://github.com/nuts-foundation/nuts-node/issues/2418
-	// TODO: verify parameters
-	return scope, nil
+	// Check VP signer == VC subject (presenter is holder of VCs)
+	if vp.Holder == nil {
+		// Is this even possible?
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "verifiable presentation is invalid: no holder",
+		}
+	}
+	err = credential.VerifyPresenterIsHolder(vp)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "verifiable presentation is invalid: " + err.Error(),
+		}
+	}
+
+	// Validate the presentation submission:
+	// 1. Resolve presentation definition for the requested scope
+	// 2. Take VCs mapped by the presentation submission
+	// 3. Match the VCs from (2) to the definition from (1). This should yield the same list of credentials.
+	//    This actually creates a new submission.
+	presentationDefinition := s.auth.PresentationDefinitions().ByScope(scope)
+	if presentationDefinition == nil {
```

`ByScope` should probably return an error